converter/src/main/java/gov/cms/qpp/conversion/decode/QrdaDecoderEngine.java

```diff
@@ -0,0 +1,284 @@
+	 */
+	private DecodeData decodeTree(final Element element, final Node parentNode) {
+		DecodeData result = decodeSingleElement(element, parentNode);
+		DecodeResult decodedResult = result.getDecodeResult();
+		Node decodedNode = result.getNode();
+
+		decodedNode = decideNewParentNode(decodedNode, parentNode);
+
+		if (DecodeResult.TREE_FINISHED == decodedResult) {
+			return new DecodeData(DecodeResult.TREE_FINISHED, decodedNode);
+		} else if (DecodeResult.TREE_ESCAPED == decodedResult) {
+			return new DecodeData(DecodeResult.TREE_FINISHED, null);
+		}
+
+		return decodeChildren(element, decodedNode);
+	}
+
+	/**
+	 * Decodes the passed in element if it is a {@code templateId} and assigns it to the {@code parentNode}.
+	 *
+	 * @param element The element to decode.
+	 * @param parentNode The node add the child decoded {@link Node} to.
+	 * @return The tuple of a {@link DecodeResult} and {@link Node} that was decoded from the {@link Element}.
+	 */
+	private DecodeData decodeSingleElement(final Element element, final Node parentNode) {
+
+		if (!TEMPLATE_ID.equals(element.getName())) {
+			return new DecodeData(DecodeResult.TREE_CONTINUE, null);
+		}
+
+		TemplateId templateId = getTemplateId(element);
+		QrdaDecoder decoder = getDecoder(templateId);
+
+		if (null == decoder) {
+			return new DecodeData(DecodeResult.TREE_CONTINUE, null);
+		}
+
+		Node childNode = new Node(templateId, parentNode);
+		childNode.setDefaultNsUri(defaultNs.getURI());
+		decoder.setNamespace(element.getNamespace());
+
+		Element parentElement = element.getParentElement();
+
+		DecodeResult decodeResult = decoder.decode(parentElement, childNode);
+
+		if (decodeResult == DecodeResult.TREE_ESCAPED) {
+			return new DecodeData(DecodeResult.TREE_ESCAPED, null);
+		}
+
+		childNode.setPath(XPathHelper.getAbsolutePath(parentElement));
+		parentNode.addChildNode(childNode);
+
+		return new DecodeData(decodeResult, childNode);
+	}
+
+	/**
+	 * Iterates over all the children of the passed in {@link Element} and calls {@link #decodeTree(Element, Node)} on them.
+	 *
+	 * @param element The element who's children will be decoded.
+	 * @param parentNode The parent node
+	 * @return The tuple of a {@link DecodeResult} and {@link Node} that was decoded from the children.
+	 */
+	private DecodeData decodeChildren(final Element element, final Node parentNode) {
+
+		List<Element> childElements = element.getChildren();
+
+		DecodeData decodeData = new DecodeData(DecodeResult.TREE_CONTINUE, parentNode);
+
+		Node currentParentNode = parentNode;
+
+		for (Element childElement : childElements) {
+			DecodeData childDecodeData = decodeTree(childElement, currentParentNode);
+
+			DecodeResult childDecodeResult = childDecodeData.getDecodeResult();
+			Node childDecodedNode = childDecodeData.getNode();
+
+			if (DecodeResult.TREE_FINISHED == childDecodeResult) {
+				decodeData = new DecodeData(DecodeResult.TREE_CONTINUE, currentParentNode);
```

Comment:
The more I think about this the more I'm leaning toward the initial parent `parentNode` being the appropriate reference to return. At [this point](https://github.com/CMSgov/qpp-conversion-tool/pull/585/files#diff-9f9b0d1a1b04b8aa8fd704e20aece6daR99) do you really care which descendant node was the last decoded? I think the shared ancestor would promote the most utility. Unless I'm missing something the only benefit I see to using `currentParentNode` over `parentNode` is the added difficulty of possible refactoring the guts of the for loop.